modules/e2eping/e2eping.js

```diff
@@ -0,0 +1,333 @@
+
+                if (request.timeSent < now - this.e2eping.analyticsIntervalMs) {
+                    // An old request. We don't care about it anymore.
+                    delete this.requests[requestId];
+                } else if (request.rtt) {
+                    rtt = Math.min(rtt, request.rtt);
+                }
+            }
+        }
+
+        if (rtt < Infinity) {
+            this.sendAnalytics(rtt);
+        }
+    }.bind(this); // eslint-disable-line no-invalid-this
+
+    /**
+     * Sends an analytics event for this participant with the given RTT.
+     * @type {*}
+     */
+    sendAnalytics = function(rtt) {
+        Statistics.sendAnalytics(createE2eRttEvent(
+            this.id,
+            this.participant.getProperty('region'),
+            rtt));
+    }.bind(this); // eslint-disable-line no-invalid-this
+}
+
+/**
+ * Implements end-to-end ping (from one conference participant to another) via
+ * the jitsi-videobridge channel (either WebRTC data channel or web socket).
+ *
+ * This class:
+ * 1. Sends periodic ping requests to all other participants in the
+ * conference.
+ * 2. Responds to ping requests from other participants.
+ * 3. Fires events with the end-to-end RTT to each participant whenever a
+ * response is received.
+ * 4. Fires analytics events with the end-to-end RTT periodically.
+ */
+export default class E2ePing {
+    /**
+     * @param {JitsiConference} conference
+     * @param {Object} options
+     */
+    constructor(conference, options) {
+        this.conference = conference;
+
+        // The interval at which pings will be sent (<= 0 disables sending).
+        this.pingIntervalMs = 10000;
+
+        // The interval at which analytics events will be sent.
+        this.analyticsIntervalMs = 60000;
+
+        // Maps a participant ID to its ParticipantState
+        this.participants = {};
+
+        // Whether the WebRTC channel has been opened or not.
+        this.dataChannelOpened = false;
+
+        if (options && options.e2eping) {
+            if (typeof options.e2eping.pingInterval === 'number') {
+                this.pingIntervalMs = options.e2eping.pingInterval;
+            }
+            if (typeof options.e2eping.analyticsInterval === 'number') {
+                this.analyticsIntervalMs = options.e2eping.analyticsInterval;
+            }
+
+            // We want to report at most once a ping interval.
+            if (this.analyticsIntervalMs > 0 && this.analyticsIntervalMs
+                < this.pingIntervalMs) {
+                this.analyticsIntervalMs = this.pingIntervalMs;
+            }
+        }
+        logger.info(
+            `Initializing e2e ping; pingInterval=${
+                this.pingIntervalMs}, analyticsInterval=${
+                this.analyticsIntervalMs}.`);
+
+        // Only subscribe to user join/leave events if sending pings is enabled.
+        if (this.pingIntervalMs > 0) {
+            conference.on(
```

Comment:
i think it'd be nice to remove the need to pass the conference object to this class...what would it look like to have the creator of this class do the wiring?  i.e. have it add a handler for the conference events and attach it to the handler methods here?  and we could pass the conference's event emitter directly.  this still leaves the sendMessage method but that could just be passed as a lambda which wraps the conference method call.

Reply:
I did that, but I don't think the results are great. Let me know if you want me to keep or revert the commit.